src/main/java/de/zalando/aruha/nakadi/service/subscription/StreamingContext.java

```diff
@@ -0,0 +1,170 @@
+import de.zalando.aruha.nakadi.service.subscription.state.StartingState;
+import de.zalando.aruha.nakadi.service.subscription.state.State;
+import de.zalando.aruha.nakadi.service.subscription.zk.ZKSubscription;
+import de.zalando.aruha.nakadi.service.subscription.zk.ZkSubscriptionClient;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StreamingContext implements SubscriptionStreamer {
+    private final StreamParameters parameters;
+    private final Session session;
+    private final ZkSubscriptionClient zkClient;
+    private final KafkaClient kafkaClient;
+    private final SubscriptionOutput out;
+    private final long kafkaPollTimeout;
+
+    private final ScheduledExecutorService timer;
+    private final BlockingQueue<Runnable> taskQueue = new LinkedBlockingQueue<>();
+    private final BiFunction<Session[], Partition[], Partition[]> rebalancer;
+
+    private State currentState;
+    private ZKSubscription clientListChanges;
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreamingContext.class);
+
+    StreamingContext(
+            final SubscriptionOutput out,
+            final StreamParameters parameters,
+            final Session session,
+            final ScheduledExecutorService timer,
+            final ZkSubscriptionClient zkClient,
+            final KafkaClient kafkaClient,
+            final BiFunction<Session[], Partition[], Partition[]> rebalancer,
+            final long kafkaPollTimeout) {
+        this.out = out;
+        this.parameters = parameters;
+        this.session = session;
+        this.rebalancer = rebalancer;
+        this.timer = timer;
+        this.zkClient = zkClient;
+        this.kafkaClient = kafkaClient;
+        this.kafkaPollTimeout = kafkaPollTimeout;
+    }
+
+    public StreamParameters getParameters() {
+        return parameters;
+    }
+
+    public ZkSubscriptionClient getZkClient() {
+        return zkClient;
+    }
+
+    public String getSessionId() {
+        return session.getId();
+    }
+
+    public KafkaClient getKafkaClient() {
+        return kafkaClient;
+    }
+
+    public SubscriptionOutput getOut() {
+        return out;
+    }
+
+    public long getKafkaPollTimeout() {
+        return kafkaPollTimeout;
+    }
+
+    @Override
+    public void stream() throws InterruptedException {
+        currentState = new State() {
+            @Override
+            public void onEnter() {
+            }
+        };
+
+        // Add first task - switch to starting state.
+        switchState(new StartingState());
+
+        while (null != currentState) {
+            final Runnable task = taskQueue.poll(1, TimeUnit.MINUTES);
```

Comment:
Just wondering why the timeout is 1 minute. It waits when the queue will have some tasks from the states, right?